### ISSUE TITLE
Questions about Dualrefine

### ISSUE THREAD

**Issue by STUDYHARD2113** [antabangun/DualRefine#3]
Thank you very much for your excellent work！
I have the following questions and look forward to your clarification:

(1)In depth_pose.py
```
self.pnets = nn.ModuleList()
        for _ in range(3):
            pnets = nn.ModuleList()
```
it includes 3 posenets, but it seems that in the entire code, the last 2 posenets not be used due to the constant padding of the image with zeros (where side is always 0). 
`
            img_inputs.append(imgs[0].new_zeros(imgs[0].shape[0], dtype=int))
`
It appears that only one posenet (with side=0) is utilized. I would like to inquire about the roles and functionalities of the other two posenets.

(2)During the validation process, at line 302 in trainer.py, edge_mask/mask/max_d are used. However, it seems that during the validation process, there is no corresponding generation of edge_mask/mask/max_d within the dep_forward function for the valid_sample. Consequently, there is a possibility that the edge_mask used in the validation process corresponds to the last sample from the training data. I would like to confirm whether this issue exists. 
Your insights and guidance on this matter would be greatly appreciated.

Thanks a lot

**Comment by antabangun** [comment]
Hi, thank you very much for your interest in our work!

(1) This was just something from our past experiments when we wanted to apply posenets to 3 different cameras in a surround-view setup (front, left, right). They have no functionality here and can be removed. Some other parts of the code are also from previous experiments and can be removed.

(2) The masks are only used to compute the consistency loss during training in `self.compute_losses`. During validation, we did not use these losses as validation metrics we keep an eye on. But you are right, this is not a good way to implement the codes, and maybe I will make a fix later.

I hope this answers your questions, and correct me if there is a mistake.
Thank you.

**Comment by STUDYHARD2113** [comment]
Hi, I greatly appreciate your prompt response, and once again, thank you for sharing this fantastic work. Your answers have addressed my concerns. I would like to discuss further:
(1)During the validation process, which loss function do you primarily focus on to evaluate the performance? Is it the reprojection loss from the photometric loss?
(2)Can the use of a pre-trained DIFFNET result in faster convergence of the DEQ module?
(3)Have you experimented with training on multiple GPUs using DistributedDataParallel? When I attempted parallel training, DataParallel seemed to work, but I encountered errors with DistributedDataParallel, possibly due to the iterative computation process in the DEQ module

Thank you again for your patient responses!!

**Comment by antabangun** [comment]
(1) Yes, we just monitored the reprojection loss and made sure it consistently went down. But after reading your comment, we might have overlooked this, and the reprojection loss might have been multiplied by the mask even during validation, possibly with masks from the last sample. We will need to double-check later and push a fix if necessary. Thank you for pointing this out.
(2) I am not sure about the effect of using pre-trained module or training from scratch. We did not conduct extensive experiments on this part.
(3) Unfortunately, we conducted all experiments on a single GPU, so we can't help you with issues related to DistributedDataParallel.

**Comment by STUDYHARD2113** [comment]
Thank you for your patient responses!

**Comment by STUDYHARD2113** [comment]
Thank you very much for your excellent work！
When I use the weights you provided to test the depth estimation results at 640*192 resolution, the refined depth results are similar to those in the paper. However, the initial results show significantly higher sqrel and rmse metrics compared to the results in your paper. The results of my testing are as follows: 
                                                         abs_rel |   sq_rel |     rmse | rmse_log |       a1 |       a2 |       a3 | 
                                                        0.103  &   0.776  &   4.491  &   0.181  &   0.894  &   0.965  &   0.983 
results in your paper  DualRefine-initial (D0)：
                                                        abs_rel |   sq_rel |     rmse | rmse_log |       a1 |       a2 |       a3 | 
  0.103  &   0.721  &  4.476  &  0.180  &  0.891  &  0.965  &  0.984
What could be the issue? 
Thanks a lot！

**Comment by antabangun** [comment]
Thank you very much for your continued interest in our work! Sorry, I haven't had the chance to make the previous fix.
I believe you are seeing the results from the [arxiv version](https://arxiv.org/pdf/2304.03560.pdf)? The weights we provided are from the one we used for this version.
We uploaded the arxiv version with the model which gave us slightly better overall results (except for sq rel and d3) for the refined depths compared to the [cvpr version](https://openaccess.thecvf.com/content/CVPR2023/papers/Bangunharcana_DualRefine_Self-Supervised_Depth_and_Pose_Estimation_Through_Iterative_Epipolar_Sampling_CVPR_2023_paper.pdf). There might have been another oversight in which we didn't update the table for the initial results. 
Thank you very much again for notifying us. We will update the arxiv paper asap.
May I ask the eval results for the refined depths on your end to make sure it matches with the arxiv results?

**Comment by STUDYHARD2113** [comment]
Thank you very much for your response!
My eval results for the refined depths from the released weight:
abs_rel | sq_rel | rmse | rmse_log | a1 | a2 | a3 |
0.088 | 0.701 | 4.234 | 0.170 | 0.914 | 0.967 | 0.983
The refined results I obtained are almost identical to those in your arxiv version, but there is a significant difference in the initial results (as shown in the previous comment). Therefore, I guess that you may not have updated the initial results in the arxiv version and instead have used the results from the CVPR version. This is because the description in your paper that the initial and refined results should come from different elements of the same test weight (Initial depth ('disp', 0, 0) and Refined depth ('disp', 0, 1)).

I'd like to ask what changes you made from the CVPR version to the arxiv version. Have you analyzed why there is such a noticeable decrease in sqrel (although it is still good)? 

Thank you very much!

**Comment by antabangun** [comment]
For the arxiv version, we only trained the models for 15 epochs instead of the 15+5 we did for the cvpr version. We decided on this because we observed little difference in the results when we trained more after freezing the teacher models. 

We also added some more ablation experiments with the consistency masks (table 2) after the cvpr submission. We observed the model with (Wq/Wh,k/T*) setting gave slightly better overall eval results. Unfortunately, I cannot really say why there is a drop in the sq rel as it is difficult to analyze, but I hope this answer helps.
